### DISCUSSION TITLE
OpenglES in maplibre

### DISCUSSION THREAD

**Discussion by zjl114012** [maplibre/maplibre-native#1984] [Q&A] [answered]
If I want to use Opengles to render my own things onto the bottom map of the maplibre, how should I do it? Is there a simple example

**Comment by louwers** [comment]
Hi @zjl114012, please check out the [custom layer functionality](https://github.com/maplibre/maplibre-native/blob/main/test/api/custom_layer.test.cpp).

Work is underway to replace the OpenGL ES specific custom layer functionality with something platform-agnostic (that will work for both OpenGL ES and Metal) but for now this is what you should use if you want to render something with full control.

**Reply by zjl114012** [reply]
Okay, thank you very much. This part is a combination of OpenGL and MapLibre. I will try to convert it to the combination of OpenGLE and MapLibre

**Reply by zjl114012** [reply]
Your answer is that you are currently researching how to render Opengles onto the map base, right? I don't want to render in a completely controlled manner, I just want to load the points I drew with the tile shader onto the map

**Reply by louwers** [reply]
@zjl114012 No that is possible already. Please see the example (test) that I linked. There is also an example for iOS and Android. I don't know which platform you are using.

What are you trying to do exactly? Most use cases should not require you to use a custom layer.

**Reply by zjl114012** [reply]
I want to use this platform to load many points onto the image, in the form of dots, to facilitate the display of a feature I have created. Currently, the interface seems to only provide adding marked points and JSON formatted points.

**Reply by zjl114012** [reply]
If Opengles renders the layers themselves, the speed will be relatively faster, and the speed of adding marked points one by one will be slow.

**Reply by zjl114012** [reply]
I am on the Android platform

**Reply by zjl114012** [reply]
Okay, I am compiling the mbgl test runner on Android devices for test, and I don't seem to see the shader code for Opengles in this section of the code

**Reply by zjl114012** [reply]
Sorry, I have been unable to compile mbgl test runner on Android devices with the following error. Besides, I have looked at the entire project and it seems that there is no code related to OpenGLES?
ERROR：
Could not create task ':app:processDebugResources'.
Cannot use @TaskAction annotation on method IncrementalTask.taskAction$gradle_core() because interface org.gradle.api.tasks.incremental.IncrementalTaskInputs is not a valid parameter to an action method.

**Reply by louwers** [reply, marked as answer]
Here are the relevant Android files that showcase how to use a custom layer.

https://github.com/maplibre/maplibre-native/blob/47028a62e2afc488f608656268c7d43f451e3abe/platform/android/src/example_custom_layer.cpp#L125

https://github.com/maplibre/maplibre-native/blob/47028a62e2afc488f608656268c7d43f451e3abe/platform/android/MapboxGLAndroidSDKTestApp/src/main/java/org/maplibre/android/testapp/activity/customlayer/CustomLayerActivity.kt#L21

**Reply by louwers** [reply]
@zjl114012 Make sure you are using the Gradle wrapper, so you are using the correct Gradle version. That error is not releated to OpenGL ES.

**Reply by zjl114012** [reply]
I tried to compile MapboxGLAndroidSDKTestApp under Linux, but when using make android configuration in the root directory, the following error was displayed. I would like to ask if this is my own problem, or can you help me solve this problem.
Sed: can't read tests/skipped. txt: No such file or directory
Sed: can't read tests/skipped. txt: No such file or directory
Sed: can't read tests/skipped. txt: No such file or directory
Sed: can't read tests/skipped. txt: No such file or directory
Make: * * * No rule to make target 'Android configuration' Stop

**Reply by zjl114012** [reply]
I am trying to compile, and currently many versions are not compatible and are still being downloaded. May I ask if there is a simple use case like Getting Started.

**Reply by louwers** [reply]
You do not need to compile the library yourself as an end user, you can use the published library. There is a Getting Started guide for this [here](https://maplibre.org/maplibre-native/docs/book/android/getting-started-guide.html).

You can use [this guide](https://github.com/maplibre/maplibre-native/blob/main/platform/android/DEVELOPING.md#getting-the-source) to get started with Android development of the library itself.

**Reply by zjl114012** [reply]
Yes, I tried and now I can successfully obtain the base image or use MapboxGLSurfaceView to render my own shape, such as a triangle. However, I am still not very clear about how to add these two together. The example of Test activity showcasing the Custom Layer API mentioned earlier was written in C++, should I also follow this example. I have attached the code written in Java. 
[myapplication.zip](https://github.com/maplibre/maplibre-native/files/13771140/myapplication.zip)

**Reply by zjl114012** [reply]
Sorry to bother you, is the latest version of Android Studio necessary? I encountered the following error while compiling in Android Studio:

FAILURE: Build failed with an exception
*What wet wrong:
Could not create task ': MapboxGLAndroidSDK: generateLegacyDebugRFile'
Cannot use @ TaskAction annotation on method IncrementalTask. taskAction $gradle_ Core() because interface org. gradle. app. tasks. incremental. IncrementalTaskInputs is not a valid parameter to an action method

**Reply by zjl114012** [reply]
I'm very sorry, I encountered a new problem while compiling:
[CXX1429] error when building with cmake using/home/zjl/Desktop/android/MapboxGLAndroidSDK/src/cpp/CMakeLists. txt: -- The C compiler identification is Clang 14.0.6
--The CXX compiler identification is Clang 14.0.6

**Reply by zjl114012** [reply]
I have resolved the compilation issue before, and I am very sorry for the error caused by the incomplete code I downloaded myself.

**Reply by zjl114012** [reply]
Hello, I have a new question now. The coordinates of the points I input are longitude and latitude. When I want to render them to the map through Opengl, I should need a projection. I am in the drawable_ Custom_ Layer_ Host_ Twicker.cpp found it
Mat4 projMatrix;
State. getProjMatrix (projMatrix);
Parameters. projectionMatrix=projMatrix;
Then input gl into the opengl tile shader_ Position=u_ Matrix * vec4 (a_pos, 0, 1); But this doesn't seem to be the projection transformation matrix I need either.
